test/kb_Metrics_server_test.py

```diff
@@ -4,12 +4,13 @@
 import json  # noqa: F401
```

Comment:
I won't go over every of your test logic here. But I can see 2 code errors with my bare eyes. Tests are supposed to catch them. And you only tested base cases for func. For instance, there are at least 10 if_else in _join_task_ujs. You only test 1 case. It's up to you if you want to fix your tests.

Reply:
I completely agree.  Same for _join_task_ujs, for the 1 test case, it's already 100 lines.  I will add more later.

Reply:
I'd like to merge this rounds of refactored changes to the master and then register in CI.  I will add more tests on each current and new function/feature as the development goes on.